### PR TITLE
Image upload Service

### PR DESCRIPTION
**Pull request by carlosbflores**
Image Upload Service implementation, for uploading images to AWS S3 service for Organizations.
Change on Organization Service, DTO, and Entity, for adding organizations with logo image.
DB changelog, add column logo_url into organizations table.

### CODE REVIEW

File: src/main/java/py/org/fundacionparaguaya/pspserver/config/ApplicationProperties.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    public String getFolderPath() {
+        return folderPath;
+    }
+
+    public String getFileNamePrefix() {
+        return fileNamePrefix;
+    }
+}
```
Comment (rodrigojv) [inline src/main/java/py/org/fundacionparaguaya/pspserver/config/ApplicationProperties.java]: If possible, we should avoid `@Value` annotations and do it instead by naming convention. Example:
```java
@ConfigurationProperties("application")
public class ApplicationProperties {

	private Aws aws = new Aws();
	
	public Aws getAws() {
		return this.app;
	}

	public void setAws(App aws) {
		this.aws= aws;
	}
	
	public static class Aws {
		private String accessKeyID;
```

And then in the properties:
```
application.aws.accessKeyID=${AWS_ACCESS_KEY}
```

File: src/main/java/py/org/fundacionparaguaya/pspserver/system/services/impl/ImageUploadServiceImpl.java [hunk 2]
```diff
@@ -22,28 +22,16 @@
...
-
-    @Value("${aws.bucketName}")
-    String bucketName;
-
-    @Value("${aws.folderPath}")
-    String folderPath;
-
-    @Value("${aws.fileNamePrefix}")
-    String fileNamePrefix;
+    ApplicationProperties applicationProperties;
```
Comment (rodrigojv) [inline src/main/java/py/org/fundacionparaguaya/pspserver/system/services/impl/ImageUploadServiceImpl.java]: Preferably `private final ApplicationProperties applicationProperties`.

File: src/main/java/py/org/fundacionparaguaya/pspserver/network/services/impl/OrganizationServiceImpl.java [hunk 3]
```diff
@@ -40,10 +40,12 @@
 	
 	 private final OrganizationMapper organizationMapper;
 
-	@Autowired
-	private ImageUploadService imageUploadService;
+	 @Autowired
+	 private ImageUploadService imageUploadService;
```
Comment (rodrigojv) [inline src/main/java/py/org/fundacionparaguaya/pspserver/network/services/impl/OrganizationServiceImpl.java]: `@Autowired` annotation not necessary and should be `private final`.

Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/network/services/impl/OrganizationServiceImpl.java]: `@Autowired` removed, Service setted to `private final`, setted in constructor.

File: src/main/java/py/org/fundacionparaguaya/pspserver/config/ApplicationProperties.java (hunk 1 above)
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/config/ApplicationProperties.java]: Changed `@Value` annotation by `@ConfigurationProperties` prefix and naming conventions.

File: src/main/java/py/org/fundacionparaguaya/pspserver/system/services/impl/ImageUploadServiceImpl.java (hunk 2 above)
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/system/services/impl/ImageUploadServiceImpl.java]: Changed to `private final`

File: src/main/java/py/org/fundacionparaguaya/pspserver/system/services/impl/ImageUploadServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Override
+    public String uploadImage(String fileString, Long entityId) throws IOException {
+
+        String url = null;
+
+        // Image format validation (image/jpg, image/jpeg, image/png, ...)
+        String data = fileString.substring(0, fileString.indexOf(';'));
+        String attribute = data.substring(0, data.indexOf(':'));
```
Comment (rodrigojv) [inline src/main/java/py/org/fundacionparaguaya/pspserver/system/services/impl/ImageUploadServiceImpl.java]: We can try to put all logic related to the image string parsing in a separate class.
Expected API:

```java
ImageDTO image =ImageParser.of(fileString)
  .parse();
 if (image == null ) {
   throw new IllegalArgumentException("invalid image" );
}
   // we have a valid image object
   // image.getFormat();
  uploadToAWS(image);
```
The ImageParser could look like:
```java
class ImageParser {
  public static ImageParser of(String fileString) {
    return new ImageParser(fileString);
 }
public ImageDTO parse() {
    // parse this.fileString image and if it is valid
    // retun new ImageDTO(contentType, ....));
   // else
  // return null;
}

Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/system/services/impl/ImageUploadServiceImpl.java]: Implemented ImageParser for parsing and validation of image file uploads.

File: src/main/java/py/org/fundacionparaguaya/pspserver/system/services/impl/ImageUploadServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                    String bucketName = applicationProperties.getBucketName();
+                    String fileNamePrefix = applicationProperties.getFileNamePrefix();
+                    String folderPath = applicationProperties.getFolderPath();
+                    String fileName = fileNamePrefix + entityId + "." + format;
+                    String keyName = folderPath + fileName;
+
+                    s3Client.putObject(new PutObjectRequest(bucketName, keyName, file)
+                            .withCannedAcl(CannedAccessControlList.PublicRead));
+
+                    url = "https://s3-" + s3Client.getRegionName() + ".amazonaws.com/" + bucketName + "/" + keyName;
```
[the remark this answers is not in the document]
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/system/services/impl/ImageUploadServiceImpl.java]: Building the `URL` left unchanged due to `s3Client.getURL(bucket,key)` not working as expected.

File: docs/RUN.md [hunk 6]
```diff
@@ -39,7 +40,40 @@ This will generate the initial database, and from this point any DB related chan
 Liquibase runs automatically during server startup and update the DB (if 
 needed).
 
-### Step 2, compile the application
```
Comment (rodrigojv) [inline docs/RUN.md]: Let's create a new file AWS_CONFIG.md and put this info in that file.

File: docs/RUN.md [hunk 7]
```diff
@@ -53,7 +87,7 @@ If you'd rather run unit tests:
 mvn clean install
 ```
 
-### Step 3, run the application
+### Step 5, run the application
```
Comment (rodrigojv) [inline docs/RUN.md]: At the end of this RUN.md add a link to the AWS_CONFIG.md file.

File: src/main/java/py/org/fundacionparaguaya/pspserver/network/services/impl/OrganizationServiceImpl.java [hunk 8]
```diff
@@ -95,8 +97,8 @@ public OrganizationDTO addOrganization(OrganizationDTO organizationDTO) throws I
 		OrganizationEntity newOrganization= organizationRepository.save(organization);
 
 		// Upload image to AWS S3 service
-		String file = organizationDTO.getFile();
-		String logoURL = imageUploadService.uploadImage(file, newOrganization.getId());
+		ImageDTO image = ImageParser.parse(organizationDTO.getFile());
```
Comment (rodrigojv) [inline src/main/java/py/org/fundacionparaguaya/pspserver/network/services/impl/OrganizationServiceImpl.java]: When constructing the image dto, we should provide the target directory and the image prefix.
```java
ImageDTO image = ImageParser.parse(file, appProperties.getOrgConfig());
// image has a targetDirectory property and targetImagePrefix
```

File: src/main/java/py/org/fundacionparaguaya/pspserver/network/services/impl/OrganizationServiceImpl.java [hunk 9]
```diff
@@ -95,8 +97,8 @@ public OrganizationDTO addOrganization(OrganizationDTO organizationDTO) throws I
...
+		String logoURL = imageUploadService.uploadImage(image, "organization", newOrganization.getId());
```
Comment (rodrigojv) [inline src/main/java/py/org/fundacionparaguaya/pspserver/network/services/impl/OrganizationServiceImpl.java]: Remove the second parameter.

File: src/main/java/py/org/fundacionparaguaya/pspserver/system/services/impl/ImageUploadServiceImpl.java [hunk 10]
```diff
@@ -46,10 +42,15 @@ public String uploadImage(String fileString, Long entityId) throws IOException {
...
-                String fileName = fileNamePrefix + entityId + "." + image.getFormat();
-                String keyName = folderPath + fileName;
+
+                String imageDirectory = null;
+                String imageNamePrefix = null;
+                if (entity.equals("organization")) {
+                    imageDirectory = applicationProperties.getAws().getOrgsImageDirectory();
+                    imageNamePrefix = applicationProperties.getAws().getOrgsImageNamePrefix();
+                }
+                String fileName = imageNamePrefix + entityId + "." + image.getFormat();
```
Comment (rodrigojv) [inline src/main/java/py/org/fundacionparaguaya/pspserver/system/services/impl/ImageUploadServiceImpl.java]: We can use here `image.getTargetDirectory`, same for the prefix. Removing the second parameter and the if statement.

**Comment by rodrigojv** [comment]
We have some conflicts @carlosbflores .

File: docs/RUN.md (hunk 6 above)
Comment (carlosbflores) [reply docs/RUN.md]: Added AWS_CONFIG.md file with step to set up AWS.

File: docs/RUN.md (hunk 7 above)
Comment (carlosbflores) [reply docs/RUN.md]: AWS_CONFIG linked in RUN.md

File: src/main/java/py/org/fundacionparaguaya/pspserver/network/services/impl/OrganizationServiceImpl.java (hunk 8 above)
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/network/services/impl/OrganizationServiceImpl.java]: ImageUploadService upload method simplified, imageDirectory and imageNamePrefix fields moved inside ImageDTO.

File: src/main/java/py/org/fundacionparaguaya/pspserver/network/services/impl/OrganizationServiceImpl.java (hunk 9 above)
Comment (carlosbflores) [reply src/main/java/py/org/fundacionparaguaya/pspserver/network/services/impl/OrganizationServiceImpl.java]: Parameter `entity` removed, imageDirectory and imageNamePrefix obtained directly from ImageDTO.

**Comment by carlosbflores** [comment]
I'll check the conflicts with branch develop now, I'll push when resolved.

**Comment by gmartinezmarin** [comment]
Issue of PMD that has to be resolved

**Pull request merged**